File: 프로그래머스/lv2/42883. 큰 수 만들기/큰 수 만들기.py
from collections import deque
def solution(number, k):
    number = deque(number)
    q = deque()
    flag = 0
    while k>0 and len(number)>0 :
        if len(q) == 0:
            q.append(number.popleft())
        if q[-1]<number[0]:
            flag = 1
            q.pop()
            k-=1
        else:
            q.append(number.popleft())
    if flag == 0:
        for i in range(k):
            q.pop()
    return ''.join(list(q)) + ''.join(list(number))

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.js
function solution(number, k) {
    const stack = [];
    let cnt = 0;
    
    for (let i = 0; i < number.length; i++) {
        const num = number[i];
        while (cnt < k && stack.length > 0 && stack[stack.length - 1] < num) {
            stack.pop();
            cnt++;
        }
        stack.push(num);
    }
    while (cnt < k && stack.length > 0) {
        stack.pop();
        cnt++;
    }
    
    return stack.join("");
}